federation/planner.go

```
@@ -0,0 +1,300 @@
+		var childPlan *Plan
+		if selection.SelectionSet != nil {
+			// XXX: assert existence of types elsewhere?
+			var err error
+			// XXX type assertoin
+			childPlan, err = e.plan(field.Type, selection.SelectionSet, service)
+			if err != nil {
+				return nil, fmt.Errorf("planning for %s: %v", selection.Name, err)
+			}
+		}
+
+		// newSelection := &graphql.RawSelection{
+		// 	Alias: selection.Alias,
+		// 	Name:  selection.Name,
+		// 	Args:  selection.Args,
+		// }
+		// if childPlan != nil {
+		// 	newSelection.SelectionSet = childPlan.SelectionSet
+		// }
+
+		// p.SelectionSet.Selections = append(p.SelectionSet.Selections, newSelection)
+		p.SelectionSet.Selections = append(p.SelectionSet.Selections, selection)
+		if childPlan != nil {
+			for _, subPlan := range childPlan.After {
+				subPlan.Path = append(subPlan.Path, PathStep{Kind: KindField, Name: selection.Alias})
+				p.After = append(p.After, subPlan)
+			}
+		}
+	}
+
+	needKey := false
+
+	var otherServices []string
+	for other := range selectionsByService {
+		otherServices = append(otherServices, other)
+	}
+	sort.Strings(otherServices)
+
+	for _, other := range otherServices {
+		selections := selectionsByService[other]
+		needKey = true
+
+		// what if a field has multiple options? should we consider capacity?
+		// what other fields we might want to resolve after?
+		// nah, just go with default... and consider being able to stick with
+		// the same a bonus
+		subPlan, err := e.plan(typ, &graphql.SelectionSet{Selections: selections}, other)
+		if err != nil {
+			return nil, fmt.Errorf("planning for %s: %v", other, err)
+		}
+
+		// p.AfterNode.Services[other] = subPlan
+		p.After = append(p.After, subPlan)
+	}
+
+	if needKey {
+		hasKey := false
+
+		for _, selection := range p.SelectionSet.Selections {
+			if selection.Name == "__federation" && selection.Alias == "__federation" {
+				hasKey = true
+			} else if selection.Alias == "__federation" {
+				// error, conflict, can't do this.
```

let's return an error here?